### PR TITLE
Add the library search box on the 404 page

### LINKED ISSUE #14572: Doc landing page and the 404 page could have the library search box
http://doc.rust-lang.org/master/ and http://doc.rust-lang.org/master/doesntexist could hook into the javascript that runs the search on http://doc.rust-lang.org/master/std/ to all show a search box themselves.

(The 404 page could even be intelligently auto-filled based on the URL, as an additional enhancement.)

### PR DESCRIPTION
**Pull request by kzys**
It actually has a link to search already, but it would be better to
have the search "box" as like index.md to be consistent.

<style> can be shared with index.md, but these pages currently
use https://doc.rust-lang.org/rust.css directly.

Fixes #14572.

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
Thanks! Can you provide a screenshot as well please? :)

**Comment by GuillaumeGomez** [comment]
The UI is a bit weird... Having two fields following each other isn't really nice. What about a dropdown menu on the left of the input field where you can pick either you want to search on duckduckgo or in the std docs?

**Comment by kzys** [comment]
How about having 2 buttons instead of the dropdown?

```
[not_found    ][Search Standard Library][Search the Internet]
```

or

```
[not_found    ]
[Search Standard Library][Search the Internet]
```

Having a dropdown menu makes sense when there are a lot of options, but we only have two here. Normal buttons allow users to see all of them at once and click one of them, while dropdowns require multiple clicks to perform search.

**Comment by GuillaumeGomez** [comment]
But then by default, if the user press "enter", which button should be used?

**Comment by kzys** [comment]
Standard Library? It would be great if we have some data from
doc.rust-lang.org, but I think we don't have that.

Google is using this "left button as default" on normal search vs. I'm
feeling lucky. People would see the left button as default, I assume.

On Sep 28, 2018 5:28 AM, "Guillaume Gomez" <notifications@github.com> wrote:

But then by default, if the user press "enter", which button should be used?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/rust-lang/rust/pull/54609#issuecomment-425419521>, or mute
the thread

**Comment by kzys** [comment]
How about this one? The default is clear. All the options are shown. You still need multiple-click to search doc.rust-lang.org though.

untitled

**Comment by GuillaumeGomez** [comment]
No more duckduckgo?

**Comment by kzys** [comment]
"From doc.rust-lang.org" uses DuckDuckGo. I've uploaded the files on my S3 bucket. Can you take a look?

http://rust-pull-request.s3-website-us-east-1.amazonaws.com/54609/not_found.html

**Comment by GuillaumeGomez** [comment]
Then please replace "From doc.rust-lang.org" with "From duckduckgo".

**Comment by kzys** [comment]
Sure. The last commit changed that to "From DuckDuckGo"

File: src/doc/not_found.md [hunk 1]
```diff
@@ -13,11 +35,20 @@ Some things that might be helpful to you though:
...
+  <form id="search-form" action="https://duckduckgo.com/">
+    <input id="search-input" type="search" name="q"></input>
+    <input type="submit" value="Search" id="search-but">
+    <!--
+      Don't show the options by default,
+      since "From the Standary Library" doesn't work without JavaScript
+    -->
+    <fieldset id="search-from" style="display:none">
+      <label><input name="from" value="library" type="radio"> From the Standard Library</label>
+      <label><input name="from" value="dro" type="radio" checked> From DuckDuckGo</label>
```
[the remark this answers is not in the document]
Comment (kzys) [reply src/doc/not_found.md]: Sure. Changed at 23af6bb.

**Comment by GuillaumeGomez** [comment]
Thanks!

@bors: r+ rollup

**Pull request merged**